### DISCUSSION TITLE
ExternalOutputInterface for all simulations

### DISCUSSION THREAD

**Discussion by pepijndevos** [Xyce/Xyce#20] [Q&A] [answered]
I'm writing a [simulation server](https://github.com/NyanCAD/XyceSimServer) that will allow running Xyce on a beefy server in a docker image and run simulations on it remotely.

I have written an ExternalOutputInterface to capture simulation output, but it requires specifying which simulation to listen for. I was wondering what would be a good way to capture all simulation output?

The only way I can see is to instantiate an interface for every possible simulation type, and after the simulation figure out which one(s) have data in them.

**Comment by tvrusso** [comment, marked as answer]
The requirement of specifying which simulation to listen for mirrors the requirement that every ".print" line specify the analysis with which it is associated.  In fact, it derives from the internals of the code that make the whole thing work.

The only way there *is* is to instantiate separate interface for each simulation type.

There is no problem with registering interfaces that go unused, just as there's no problem having .print statements that do nothing for the analysis that's actually being run (e.g. you could have a .print HB-FD line in a transient netlist, it just wouldn't *do* anything).

I can envision may ways to have your ExternalOutputInterface implementations signal to your main program that they were actually used.  For example, they could be set up to call functions of your own when they are actually called, or be augmented to have accessor functions to check if they were actually used.  Xyce won't care what else the object does so long as you implement the set of functions required for the interface.

**Reply by tvrusso** [reply]
I said:  The requirement of specifying which simulation to listen for mirrors the requirement that every ".print" line specify the analysis with which it is associated. In fact, it derives from the internals of the code that make the whole thing work.

But what I should have emphasized is that the external I/O system is really just a way of bypassing Xyce's parsing of .print lines and going directly to the creation of an outputter --- so just about everything associated with the external output system is designed to confuse Xyce into thinking it got a .print line somehow, but to call your functions for any output associated with that line instead of outputting to a file in a given format.

There is no workaround to make Xyce call the same interface object for *every* analysis --- each object must tell Xyce what analysis it applies to, and it can only apply to one.  You could always create a sort of mega output class derived from the ExternalOutputInterface interface class and have it so that you can configure that object's behavior using your own methods, and then pass Xyce several instances of that mega object, each that returns a different value for getOutputType().

How you actually do that is where you'll have to get creative.
